### PR TITLE
initial state machine

### PR DESCRIPTION
**Pull request by josepvalls**

### CODE REVIEW

File: scripts/components/Stateful.gd [hunk 1]
```diff
@@ -0,0 +1,47 @@
+class_name StatefulComponent
+extends BaseComponent
+
+
+const COMPONENT_NAME := &"StatefulComponent"
+
+func get_component_name() -> StringName:
+	return COMPONENT_NAME
+
+var current_state: String = ""
```
Comment (Jagholin) [inline scripts/components/Stateful.gd]: can just be of type `State`, so you don't have to perform a dictionary lookup all the time

File: scripts/components/Stateful.gd [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+var state_to_state_map = {}
+var state_list = []
```
Comment (Jagholin) [inline scripts/components/Stateful.gd]: these variables do basically the same thing

File: scripts/components/Stateful.gd [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+			change_state(child.name)
+
+# Change to the new state by first calling any exit logic on the current state.
+func change_state(new_state: String) -> void:
+	if current_state:
+		state_to_state_map[current_state].exit()
+	
+	current_state = new_state
+	var state = state_to_state_map[current_state]
+	state.elapsed=0
```
Comment (Jagholin) [inline scripts/components/Stateful.gd]: `State.enter()` already does this

File: scripts/components/Stateful.gd (hunk 1 above)
Comment (josepvalls) [reply scripts/components/Stateful.gd]: That was the goal. I was having trouble using the states as is so I was using a string with the state name insted. I had this map to lookup from state names to state classes themselves. I'll check this out later.

File: scripts/components/Stateful.gd (hunk 2 above)
Comment (josepvalls) [reply scripts/components/Stateful.gd]: I was planning on using those to transition between states automatically so a state doesn't need to explicitly say which state comes next. I'll revisit this idea tonight.

**Comment by josepvalls** [comment]
Thanks!

**Pull request merged**